source/language/scope.rst

```diff
@@ -0,0 +1,318 @@
+      //  - 'c': type 'uint', from the parameter list (shadows the outer 'const
+      //      int' 'c').
+      //  - 'd': type 'const int', value 4, visible from the global scope
+      //      because it is a 'const' type.
+      //  - 'local': type 'int', value 5, from regular definition in the current
+      //      scope.
+      //  - other built-in identifiers (such as 'U' and 'pi') that are
+      //      implicitly defined in the global scope.
+      //  - all available hardware qubits (such as '$0')
+      //
+      // The variable 'b' is not in scope, because its visibility as a
+      // non-'const' type does not make it available within subroutines.  The
+      // hardware qubit identifiers are in scope, but not the virtual qubit
+      // identifier 'all_qubits'.
+   }
+
+   // After the subroutine block, 'a' and 'c' once again refer to the variables
+   // of type 'int' and 'const int' defined on lines 3 and 5 respectively.
+   // 'local' (from the subroutine body) is not in scope, while 'my_routine'
+   // (the subroutine) is.
+
+   float[64] new_variable = 1.5;
+
+   def second_subroutine(qubit[4] q) {
+      int local = 8;
+
+      let alias = q[0:2];
+
+      // Identifiers in scope are:
+      //   - 'second_subroutine'
+      //   - 'my_subroutine'
+      //   - 'local': type 'int', value 8
+      //   - 'c': type 'const int', value 3
+      //   - 'd': type 'const int', value 4
+      //   - 'q': type 'qubit[4]', a virtual, run-time-known qubit register.
+      //   - 'alias': alias for the first three qubits of 'q'.
+      //   - 'new_variable': type 'float[64]', value 1.5
+      //   - the other built-in identifiers like 'U' and 'pi'
+      //   - the available hardware qubits like '$0'.
+   }
+
+
+.. _scope-block:
+
+Block scope
+-----------
+
+Certain control-flow operations introduce their own local scope.  These
+operations are:
+
+* ``for`` loops,
+* ``while`` loops,
+* ``if`` and ``else`` blocks,
+* ``box`` statements.
+
+These scopes inherit *all* variables that are in scope in the immediately
+containing scope.  Unlike subroutines and gate scopes, this includes variables
+that are not ``const``.  This is broadly similar how these constructs behave in
+other procedural languages, such as C.
+
+The iteration variable of a ``for`` loop has lifetime and visibility as if it
+were declared as the first statement in the body of that loop.  It is not
+accessible after the body of the loop.
+
+The blocks associated with ``if`` and its corresponding ``else`` define two
+different scopes; the variables and definitions are not shared between them.
+
+As with subroutine scopes, variables defined locally in these scopes (including
+the for-loop iteration variable) may shadow variables with the same name in
+outer scopes.  When the defining scope of a shadowing variable ends, the
+previous variable (which was shadowed) becomes accessible again.  Qubits and
+arrays cannot be declared within local block scopes, but aliases can.
```

Comment:
This is fair enough from a language-design perspective.  I think it might be good to talk to some implementors about making global variables available in function scopes before we move to do it - the way I've written it, we're a bit more conservative in the language, because we never need to worry about global-state mutation.  This was a feature of Michael's original draft, I just made it stricter, and perhaps a bit more consistent in the type system here by removing their visibility.

I think if global variables are visible within functions, there's probably not as strong an implementation reason to forbid mutating them.  We can discuss it more in the meeting on Thursday?

Reply:
I wasn't actually advocating for being able to mutate globals within subroutines. What makes me uneasy is the ability to shadow variables in non-subroutine scopes.

Reply:
Oh right, I understand.  This I think is the least surprising way of doing things.  It's necessary, for example, to allow the for-loop variable to shadow a previously-defined variable.  I can't think of any language off the top of my head with explicit variable declarations (e.g. C but not Python) that allow declarations in control-flow scopes, but don't permit shadowing.